File: r1-spike/harness/test/support/fake_runner.exs
# Fake runner for harness self-tests: echoes each request line. With
# `--corrupt-step N`, a world.run with more than N commands gets a changed response,
# so the minimal failing input has exactly N + 1 commands.
corrupt =
  case System.argv() do
    ["--corrupt-step", n] -> String.to_integer(n)
    [] -> nil
  end

Stream.repeatedly(fn -> IO.read(:stdio, :line) end)
|> Stream.take_while(&is_binary/1)
|> Enum.each(fn line ->
  line = String.trim_trailing(line, "\n")

  out =
    with n when is_integer(n) <- corrupt,
         {:ok, %{"fn" => "world.run", "commands" => cmds}} when length(cmds) > n <-
           JSON.decode(line) do
      "corrupt:" <> line
    else
      _ -> line
    end

  IO.write(out <> "\n")
end)

File: r1-spike/harness/test/support/fake_runner.py
"""Fake runner for harness self-tests: echoes each request line byte for byte.

With `--corrupt-step N`, a world.run with more than N commands gets a changed
response, so the minimal failing input has exactly N + 1 commands. `--crlf` ends
every line with \\r\\n. Raw bytes both ways, so invalid UTF-8 and \\r-terminated
lines echo intact.
"""
import json
import sys

corrupt = int(sys.argv[2]) if sys.argv[1:2] == ["--corrupt-step"] else None
ending = b"\r\n" if sys.argv[1:2] == ["--crlf"] else b"\n"
out = sys.stdout.buffer
for raw in sys.stdin.buffer:
    line = raw[:-1] if raw.endswith(b"\n") else raw
    if corrupt is not None:
        try:
            req = json.loads(line)
            if req.get("fn") == "world.run" and len(req["commands"]) > corrupt:
                line = b"corrupt:" + line
        except Exception:
            pass
    out.write(line + ending)
    out.flush()
